### PR TITLE
feature: supports link on connections

### PR DESCRIPTION
**Pull request by danielsuh05**
Connections now support `link: ` (fixes issue #1827)

Some notes:
- Every single test file changed because all of the class names changed when adding `link: ""`.
- Links don't work yet on connections in PDFs, not sure exactly how to do that. 
- Check `link-on-connections` for the test I used.

### CODE REVIEW

File: d2exporter/export.go [hunk 1]
```diff
@@ -356,7 +356,6 @@ func toConnection(edge *d2graph.Edge, theme *d2themes.Theme) d2target.Connection
 	}
 	if edge.Link != nil {
 		connection.Link = edge.Link.Value
-		connection.PrettyLink = toPrettyLink(edge.Dst.Graph, edge.Link.Value)
```
Comment (alixander) [inline d2exporter/export.go]: why no pretty link?

File: ci/release/changelogs/next.md [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 - Opacity 0 shapes no longer have a label mask which made any segment of connections going through them lower opacity [#1940](https://github.com/terrastruct/d2/pull/1940)
 - Bidirectional connections are now animated in opposite directions rather than one direction [#1939](https://github.com/terrastruct/d2/pull/1939)
+- Connections now support `link` in SVGs [#1955](https://github.com/terrastruct/d2/pull/1955)
```
Comment (alixander) [inline ci/release/changelogs/next.md:7]: this counts a new feature!

**Comment by danielsuh05** [comment]
Oh I see what you mean. I was trying to figure out how exactly you wanted it implemented so I decided to do this for now and get a review. I'll work on it

**Comment by danielsuh05** [comment]
Was this more what you were thinking?

File: d2graph/d2graph.go [hunk 3]
```diff
@@ -1549,7 +1549,11 @@ func (g *Graph) SetDimensions(mtexts []*d2target.MText, ruler *textmeasure.Ruler
 
 		obj.SizeToContent(contentBox.Width, contentBox.Height, paddingX, paddingY)
 	}
-	for _, edge := range g.Edges {
+	for i, edge := range g.Edges {
+		if edge.Link != nil {
+			g.Edges[i].Label.Value = edge.Link.Value
+		}
```
Comment (alixander) [inline d2graph/d2graph.go:1555]: this shouldn't be in this function named set dimensions. probably in compiler.go.

Also the link shouldn't replace the label if there is one. The test should show an example of where the link turns the label into a hyperlink

File: d2renderers/d2svg/d2svg.go [hunk 4]
```diff
@@ -1594,6 +1594,22 @@ func EmbedFonts(buf *bytes.Buffer, diagramHash, source string, fontFamily *d2fon
...
+		[]string{
+			`text-link`,
+		},
+		`
+.text-link {
+	fill: blue;
+}
+		
+.text-link:visited {
+	fill: #1a0dab;
```
Comment (alixander) [inline d2renderers/d2svg/d2svg.go:1609]: do we have the explicitly set this? If you leave it unset, would it just work or no? If it's required, I think just the default "purple" is best.

**Review by alixander — CHANGES_REQUESTED** [review]
Yup!

File: d2renderers/d2svg/d2svg.go (hunk 4 above)
Comment (danielsuh05) [reply d2renderers/d2svg/d2svg.go:1609]: Maybe it's just a browser issue on my end but these are my results (:visited is set to true in dev tools).

| No class styling | Only `.text-link` | Original code |
|--------|--------|--------|
| Screenshot 2024-05-31 at 00 05 50 | Screenshot 2024-05-31 at 00 06 09 | Screenshot 2024-05-31 at 00 09 33 |

File: d2compiler/compile.go [hunk 5]
```diff
@@ -762,6 +762,13 @@ func (c *compiler) compileEdge(obj *d2graph.Object, e *d2ir.Edge) {
 	if e.Primary() != nil {
 		c.compileLabel(&edge.Attributes, e)
 	}
+
+	for _, field := range e.Map_.Fields {
+		if field.Name == "link" && edge.Label.Value == "" {
+			edge.Label.Value = field.Primary_.String()
```
Comment (alixander) [inline d2compiler/compile.go]: since this relies on Primary() being non-nill, it should go in the above if clause

**Review by alixander — CHANGES_REQUESTED** [review]

**Review by alixander — CHANGES_REQUESTED** [review]
oh also i think it's a security issue if the edge label is a URL and the edge link is a URL and they are not the same URL. can you make that a compiler error and add a test for it please? (i'm pretty sure we do this for tooltips and links or something already)

File: d2compiler/compile.go [hunk 6]
```diff
@@ -820,6 +822,24 @@ func (c *compiler) compileEdgeMap(edge *d2graph.Edge, m *d2ir.Map) {
 	}
 }
 
+func (c *compiler) compileEdgeLinks(edge *d2graph.Edge, m *d2ir.Map) {
```
Comment (alixander) [inline d2compiler/compile.go:825]: i think this function should just be a switch case in the function `compile.go`'s `compileReserved`

**Review by alixander — CHANGES_REQUESTED** [review]

File: d2compiler/compile.go [hunk 7]
```diff
@@ -584,6 +584,16 @@ func (c *compiler) compileReserved(attrs *d2graph.Attributes, f *d2ir.Field) {
...
+
+		if attrs.Label.Value == "" {
+			attrs.Label.Value = attrs.Link.Value
+		} else {
+			// this works because the Label attribute is set before the Link attribute
+			u, err := url.ParseRequestURI(attrs.Label.Value)
+			if err == nil && u.Host != "" {
+				c.errorf(f.Primary_.Value, "Label cannot be set to URL when link is also set (for security)")
+			}
+		}
```
Comment (alixander) [inline d2compiler/compile.go:596]: actually here you can't tell if it's a field for an edge or a shape. 

Just after `compileReserved` here then: https://github.com/alixander/d2/blob/94d5ab5d2ec4be33e4227bc767e2e87068ce00ef/d2compiler/compile.go#L832

**Review by alixander — CHANGES_REQUESTED** [review]

**Comment by huyz** [comment]
Hi, what's the status on this PR?

**Comment by alixander** [comment]
Oops I did not realize the ball was in my court for reviewing. Sorry @danielsuh05 . I'm not sure if you've moved on. No worries if so I can take over, otherwise will leave a review ASAP

**Comment by huyz** [comment]
I hope to see this soon.  Links on connections are pretty important. They open up a lot of possibilities for letting diagrams act as jump points to, e.g., code.

**Comment by danielsuh05** [comment]
I'm ok continuing with this issue.

**Comment by huyz** [comment]
Thank you @danielsuh05 and @alixander!

File: d2compiler/compile.go [hunk 8]
```diff
@@ -762,6 +762,7 @@ func (c *compiler) compileEdge(obj *d2graph.Object, e *d2ir.Edge) {
 	if e.Primary() != nil {
 		c.compileLabel(&edge.Attributes, e)
 	}
+
```
Comment (alixander) [inline d2compiler/compile.go:765]: remove unrelated newline

File: d2compiler/compile.go [hunk 9]
```diff
@@ -830,6 +831,19 @@ func (c *compiler) compileEdgeField(edge *d2graph.Edge, f *d2ir.Field) {
...
+		}
+
+		if edge.Label.Value == "" {
+			edge.Label.Value = edge.Link.Value
+		} else {
+			u, err := url.ParseRequestURI(edge.Label.Value)
+			if err == nil && u.Host != "" {
+				c.errorf(f.Primary_.Value, "Label cannot be set to URL when link is also set (for security)")
+			}
+		}
```
Comment (alixander) [inline d2compiler/compile.go:846]: this should just be in `compileReserved` i think

File: d2compiler/compile_test.go [hunk 10]
```diff
@@ -1543,6 +1543,32 @@ x -> y: {
...
 		},
+		{
+			name:   "no_url_link_and_path_url_label_concurrently",
+			text:   `x -> y: https://google.com {link: https://not-google.com }`,
+			expErr: `d2/testdata/d2compiler/TestCompile/no_url_link_and_path_url_label_concurrently.d2:1:35: Label cannot be set to URL when link is also set (for security)`,
+		},
+		{
+			name:   "url_link_and_path_url_label_ok",
+			text:   `x -> y note: {link: https://not-google.com}`,
+			expErr: ``,
```
Comment (alixander) [inline d2compiler/compile_test.go:1554]: omit this line

**Review by alixander — CHANGES_REQUESTED** [review]
can you squash all the implementation commits into one and the test diffs into another? makes it easier to review to just see the diff for the implementation

-----

the conflicts should just be a test rerun after rebasing master

File: d2compiler/compile_test.go [hunk 11]
```diff
@@ -1543,6 +1543,31 @@ x -> y: {
...
 			},
 		},
+		{
+			name:   "no_url_link_and_path_url_label_concurrently",
+			text:   `x -> y: https://google.com {link: https://not-google.com }`,
+			expErr: `d2/testdata/d2compiler/TestCompile/no_url_link_and_path_url_label_concurrently.d2:1:35: Label cannot be set to URL when link is also set (for security)`,
+		},
+		{
+			name: "url_link_and_path_url_label_ok",
+			text: `x -> y note: {link: https://not-google.com}`,
```
Comment (alixander) [inline d2compiler/compile_test.go:1553]: the colon is not in the right place, should be `x -> y: note {...}`

File: d2compiler/compile_test.go [hunk 12]
```diff
@@ -1543,6 +1543,31 @@ x -> y: {
...
+		},
+		{
+			name: "url_link_and_path_url_label_concurrently",
+			text: `x -> y: hello world {link: https://google.com}`,
```
Comment (alixander) [inline d2compiler/compile_test.go:1557]: this is the same test as above?

File: d2renderers/d2svg/d2svg.go [hunk 13]
```diff
@@ -634,11 +634,23 @@ func drawConnection(writer io.Writer, labelMaskID string, connection d2target.Co
...
 		textEl.ClassName = fontClass
 		textEl.Style = fmt.Sprintf("text-anchor:%s;font-size:%vpx", "middle", connection.FontSize)
 		textEl.Content = RenderText(connection.Label, textEl.X, float64(connection.LabelHeight))
+
+		if connection.Link != "" {
+			textEl.ClassName += " text-underline text-link"
+
+			fmt.Fprintf(writer, `<a href="%s" xlink:href="%[1]s">`, svg.EscapeText(connection.Link))
+		} else {
+			textEl.Fill = connection.GetFontColor()
```
Comment (alixander) [inline d2renderers/d2svg/d2svg.go:646]: why is fill dependent?

**Review by alixander — CHANGES_REQUESTED** [review]

**Comment by alixander** [comment]
@danielsuh05 Let me know if you'd like me to take over

**Review by alixander — APPROVED** [review]
i rebased and made minor fixes, thank you again @danielsuh05

**Comment by bo-ku-ra** [comment]
is the font color blue only?

**Comment by alixander** [comment]
yes, since that's the universal link color. But if you want it different, please make an issue

**Pull request merged**